File: Problems/241. Different Ways to Add Parentheses.py
class Solution:
    def diffWays(self,s, memo):
        if s.isdigit():
            return [int(s)]
        res = []
        
        #check if exists in memo to reduce duplicated computation
        if s in memo:
            return memo[s]
        
        for i, c in enumerate(s):
            #if we found an operator, split it and do the recusion
            #this is the same behavior of using parenthesis
            if c in "+-*":
                front = self.diffWays(s[:i], memo)
                back = self.diffWays(s[i+1:], memo)
                #caculate all the possible combinations
                for x in front:
                    for y in back:
                        res.append(self.helper(x,y,c))
        #store in memo
        memo[s] = res
        return memo[s]
                
    def helper(self,x,y,c):             
        if c == "+":
            return x+y
        elif c == "-":
            return x-y
        elif c == "*":
            return x*y
                
    def diffWaysToCompute(self, input: str) -> List[int]:
        #intuition:
        #since we are searching all the possible way
        #try recursion
        #observation: pair will be the same as operator
        #divide and recusion at every operator
        
        if not input:
            return []
        
        ans = self.diffWays(input,{})
        
        return ans
        
                
        

File: LeetCode/Problems/C++/241. Different Ways to Add Parentheses.cpp


class Solution {
public:
    vector<int> diffWays(string s, unordered_map<string, vector<int>>& memo){
        if(all_of(s.begin(), s.end(), ::isdigit)){
            return {stoi(s)};
        }
        
        if(memo.find(s)!=memo.end()){
            return memo[s];
        }
        
        vector<int> result;
        for(int i=0;i<s.size();i++){
            if(s[i] == '+' || s[i] == '-' || s[i] == '*'){
                vector<int> left = diffWays(s.substr(0,i), memo);
                vector<int> right = diffWays(s.substr(i+1), memo);
                for(int j=0;j<left.size();j++){
                    for(int k=0;k<right.size();k++){
                        result.push_back(helper(left[j], right[k], s[i]));
                    }
                }
            }
            
        }
        memo[s] = result;
        return memo[s];
    }
    
    int helper(int l, int r, char op){
        if(op == '+')
            return l + r;
        else if(op == '-')
            return l - r;
        else if (op == '*')
            return l * r;
        else
            return NULL;
    }
    
    
    
    vector<int> diffWaysToCompute(string input) {
        if(!input.size()) return {};
        
        vector<int> result;
        unordered_map<string, vector<int>> memo;
        result = diffWays(input, memo);
        
        return result;
    }
};